app/components/calculator_component.rb

```diff
@@ -12,75 +22,98 @@ class CalculatorComponent < ViewComponent::Base
 
   def initialize(total: 0)
     @buffer = total.to_s
-    @buffering = false
-    @next_number = false
   end
 
   def add_char(event)
-    str = event.target.data[:char]
-
-    unless buffering
-      @buffer = ""
-      @buffering = true
-    end
-
-    if next_number
+    str = event.target.data[:char].dup
+    if @buffer == "0"
       @buffer = str
-      @next_number = false
     else
       @buffer << str
     end
   end
 
   def operation(event)
-    @operand_one = buffer.to_f
-    @op = event.target.data[:op].to_sym
-    @next_number = true
+    event_op = event.target.data[:op].to_sym
+    @op = event_op if overwrite_operation?(event_op)
+    # Return if there's not anything to operate on
+    return if buffer.blank?
+
+    calculate
+    next_ops(drop_decimals(buffer.to_f), event_op)
   end
 
   def change_sign
-    amt = (buffer.to_f * -1)
-    @buffer = drop_decimals(amt).to_s
+    amt = (current_value.to_f * -1)
+    change_current(drop_decimals(amt))
   end
 
   def percent
-    amt = (buffer.to_f / 100.0)
-    @buffer = drop_decimals(amt).to_s
+    amt = (current_value.to_f / 100.0)
+    change_current(drop_decimals(amt))
   end
 
   def equals
-    total =
-      case op
-      when :+
-        operand_one + buffer.to_f
-      when :-
-        operand_one - buffer.to_f
-      when :*
-        operand_one * buffer.to_f
-      when :/
-        operand_one / buffer.to_f
-      end
+    return if buffer.blank?
 
-    @buffer = drop_decimals(total)
-
-    @buffering = false
+    calculate
+    next_ops(current_value, nil)
   end
 
   def clear
-    @buffering = false
-    @op = nil
-    @operand_one = nil
-    @buffer = "0"
-    @next_number = false
+    clear_ops
+  end
+
+  def current_value
+    return operand_one if buffer.blank?
+
+    buffer
   end
 
   private
 
+  def calculate
+    return unless op && operand_one && buffer.present?
+
+    total = operand_one.send(op, buffer.to_f)
```

Comment:
I think this will allow end-users to `send` any method on this object by changing a data attribute on the element. There needs to be an allowed list of `op` values.

Reply:
Thanks! i had forgotten to think about this but this is well caught.